src/Backend/NotifyManager.vala

```diff
@@ -29,21 +29,98 @@ public class Backend.NotifyManager : Object {
     }
 
     public Gee.HashMap<string, App> apps { get; construct; } /* string: app-id */
+    public Gee.HashSet<string> apps_that_have_notified { get; construct; }
 
     public string selected_app_id { get; set; }
 
     construct {
         apps = new Gee.HashMap<string, App> ();
+        var apps_that_have_notified = new Gee.HashMap<string, string> ();
 
-        var installed_apps = AppInfo.get_all ();
+        var settings = new Settings.full (
+            GLib.SettingsSchemaSource.get_default ().lookup ("io.elementary.notifications", false),
+            null,
+            null
+        );
+        foreach (var app_id in settings.get_strv("applications")) {
+            apps_that_have_notified.set(App.normalize_app_id(app_id), app_id);
+        }
+
+        foreach (AppInfo app_info in AppInfo.get_all ()) {
+            var desktop_app_info = app_info as DesktopAppInfo;
+            if (desktop_app_info == null) { continue; }
 
-        foreach (AppInfo app_info in installed_apps) {
-            DesktopAppInfo? desktop_app_info = app_info as DesktopAppInfo;
+            var app = new App(desktop_app_info);
+
+            var alternate_names = new Gee.HashMap<string, string> ();
+            // Even though it is called `alternate_names`, it includes the
+            // app's formal name as well.
+            alternate_names.set(App.normalize_app_id(app.app_id), app.app_id);
+            foreach (var alt_name in desktop_app_info.get_string_list("X-Flatpak-RenamedFrom")) {
+                alternate_names.set(App.normalize_app_id(alt_name), alt_name);
+            }
 
-            if (desktop_app_info != null && desktop_app_info.get_boolean ("X-GNOME-UsesNotifications")) {
-                var app = new App (desktop_app_info);
+            if (desktop_app_info.get_boolean ("X-GNOME-UsesNotifications")) {
+                apps_that_have_notified.unset(app.normalized_app_id);
                 apps.@set (app.app_id, app);
+            } else if (apps_that_have_notified.has_key(app.normalized_app_id)) {
+                // Notifications in this category don't necessarily declare
+                // that they send notifications in their desktop file, but
+                // we've recorded them sending at least one notification.
+
+                // We use the desktop entry given in the application's
+                // notification hint (instead of the desktop file's id) because
+                // some apps use a different formal name than the name that
+                // they send notifications as. For example, 'firefox' sends
+                // notifications belonging to 'Firefox' (note the case
+                // difference).
+                var desktop_entry = apps_that_have_notified.get(app.normalized_app_id);
+                apps_that_have_notified.unset(app.normalized_app_id);
+                apps.@set (desktop_entry, app);
+            } else {
+                foreach (var entry in alternate_names.entries) {
```

Comment:
I might have forgotten to break early out of this loop if we match an alternate name.